src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,487 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "QubitManager.hpp"
+#include "QirRuntime.hpp" // For quantum__rt__fail_cstr
+#include <cstring> // For memcpy
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+//
+// Failing in case of errors
+//
+
+static void FailNow(const char* message)
+{
+     quantum__rt__fail_cstr(message);
+}
+
+static void FailIf(bool condition, const char* message)
+{
+    if (condition)
+    {
+        quantum__rt__fail_cstr(message);
+    }
+}
+
+//
+// QubitListInSharedArray
+//
+
+QubitListInSharedArray::QubitListInSharedArray(QubitIdType startId, QubitIdType endId, QubitIdType* sharedQubitStatusArray)
+{
+    FailIf(startId > endId || startId < 0 || endId == std::numeric_limits<QubitIdType>::max(),
+        "Incorrect boundaries in the linked list initialization.");
+
```

Comment:
(I'm trying to dive deeper into the logic this time)

Does this occurrence of `std::numeric_limits<QubitIdType>::max()` mean `NoneMarker` or `MaximumQubitCapacity` or other? Consider replacing with the corresponding one, to be more informative.

Reply:
Agreed, MaximumQubitCapacity is a better id here. Potential qubit id range is from 0 to MaximumQubitCapacity - 1.